Adds the V1 Version of the DSL

VincenzoFerme opened the pull request.

jeppe-style committed:
work on test DSL example

jeppe-style committed:
updated after talk

jeppe-style committed:
Updates benchflow load test definition to DSL v1 for end-to-end
testing. Creates resource folder structure for future
benchflow-test.yml files.

jeppe-style committed:
Updates benchflow load test definition according to DSL definition.

jeppe-style committed:
Adds BenchFlowTestYamlProtocol for review by Igor.

Adds work in progress for discussion.

jeppe-style committed:
Adds implementation for deserializing and serializing of a BenchFlow
Test Definition YAML file. Tested to work with load test.

jeppe-style committed:
Enables JUnit tests to be run with maven.

jeppe-style committed:
Adds link to JUnit 4 in scala.

VincenzoFerme commented:
@igormoreno Refer to Issue #59 to start creating the BenchFlow Codacy configuration file, starting from the projects we currently have that are mainly Scala and Java. You can update this pull request with the configuration file, so it starts applying to the code committed here, and it is going to be applied to all the current and new projects that have to be updated accordingly.

VincenzoFerme commented on benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/simone/BenchFlowExperiment.scala:
```diff
@@ -0,0 +1,40 @@
+                               sutConfiguration: SutConfiguration)
+{
+  def getAliasForService(serviceName: String) = sutConfiguration.deployment.get(serviceName)
+  // TODO - move me
```
Without these methods, the DSL is not useful in BenchFlow

jeppe-style replied:
If we need helper methods I think that is fine, however, we can get it also directly from the BenchFlowTest/Experiment instance.

VincenzoFerme replied:
It is fine to have them in the experiment

VincenzoFerme commented on benchflow-dsl/src/test/scala/cloud/benchflow/dsl/definition/types/time/TimeTest.scala:
```diff
@@ -0,0 +1,42 @@
+  * @author Jesper Findahl (jesper.findahl@usi.ch) 
+  *         created on 14.03.17.
+  */
+class TimeTest extends AssertionsForJUnit {
```
Missing `micro` and check that you include all

jeppe-style replied:
Updated in commit https://github.com/benchflow/benchflow/pull/67/commits/d1b3f4c4ca1357fbcc2031a3b776a86ca1ddf240

VincenzoFerme commented:
Remaining issues:

- semantic validation https://github.com/benchflow/benchflow/issues/48
- cleanup Simone's code, and what to be taken from it https://github.com/benchflow/benchflow/issues/103
- not defined elements (e.g., strategy) https://github.com/benchflow/benchflow/issues/104
- define defaults value, when feasible and makes sense https://github.com/benchflow/benchflow/issues/105
- which types of tests we decided to implement, and what is missing here https://github.com/benchflow/benchflow/issues/106

and reply to a comment to this pull request to link them as missing here. 


Moreover:

- How can I get an experiment representation? https://github.com/benchflow/benchflow/issues/107
- Where are the different mixes definitions that were in https://github.com/benchflow/benchflow/pull/67/files#diff-759278330151bb082f8a28c58c208570

jeppe-style commented:
Mixes are defined here: https://github.com/benchflow/benchflow/tree/feature-dsl-v1-initial/benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/workload/mix

issue regarding experiment representation: https://github.com/benchflow/benchflow/issues/107

other issues:
- semantic validation https://github.com/benchflow/benchflow/issues/48
- cleanup Simone's code, and what to be taken from it https://github.com/benchflow/benchflow/issues/103
- not defined elements (e.g., strategy) https://github.com/benchflow/benchflow/issues/104
- define defaults value, when feasible and makes sense https://github.com/benchflow/benchflow/issues/105
- which types of tests we decided to implement, and what is missing here https://github.com/benchflow/benchflow/issues/106

jeppe-style committed:
Updates BenchFlowTestYamlProtocol with improved code example for the write method. Fixes superclass inheritance for TimeTest.

Merge branch 'devel' of https://github.com/benchflow/benchflow into feature-test-pull

* 'devel' of https://github.com/benchflow/benchflow:
  Adds tests folder with DSL examples.
  Update README.md
  Update README.md
  Update README.md

jeppe-style committed:
Updates TimeTest to include ‘micros’.

igormoreno reviewed (approved):
I understand that for a 1st version of the DSL a lot off "set up"/"architecture" code is necessary but I would advise that future pull requests should be much smaller and focused around a given subject.

I already talked to @jeppe-style about how to improve the tests and a good way to replace the current uses of "var" and the current calls to "get" on optional types (Try and Option). I think these changes should be incorporated in a future pull request which should also get rid of the current uses of "???".

igormoreno committed:
Scalastyle configuration file

VincenzoFerme commented:
@igormoreno I agree with you, i told the same to @jeppe-style because of the difficulties to review such big pull requests. For this pull request we agreed it is ok.

@jeppe-style can you link to the issues discussing what @igormoreno pointed out in https://github.com/benchflow/benchflow/pull/67#pullrequestreview-28423957

VincenzoFerme commented:
@igormoreno I had a look at the `scalastyle` configuration file you added. We also need https://support.codacy.com/hc/en-us/articles/115002130625-Codacy-Configuration-File/, we need to document the settings. After we verify it works, we can also add other Scala checkers configuration as well as configuring the Java tools.

igormoreno committed:
Carefully crafted changes to scalastyle config file

I took parts from the automatically generated standard scalastyle
configuration file and parts from the file used by codacy
(https://github.com/codacy/codacy-scalastyle/blob/master/src/main/resources/docs/scalastyle_config.xml)
and then made my own changes on top of it. These changes were:

- Currently there is no standard header in the files (which typically
  would contain information regarding license) so I turned that off.
- The check on naming standards for fields apparently doesn't
  distinguish real fields (on classes or traits) from constants (on
  objects) so I removed it
- I turned off the complaints about imports using wildcards because I
  think there are valid uses for it. For example when importing classes
  whose purpose is to do implicit conversions or for example when
  importing type members of a class that you are already importing (for
  example in the code I have with actors). I think this is used all over
  the scala standard library anyway
- I turned off complaints regarding grouped imports because I think that
  conflicts in some cases with having wildcard imports. Basically you
  cannot turn that on in the last option off
- I turned off complaints about imports that are not in the beginning of
  the file. I think it is important to allow that because it can bring
  benefits in terms of performance and in terms of clarity or better
  understanding of the code for example when you are importing a mutable
  collection or when you have more than one top level class (or trait or
  object) in the same file (or even if they are inside the scope of
  another class) in the bar effectively different worlds which access
  different sets of imports. This kind of code can be seen when dealing
  with actors for example.

igormoreno committed:
scalastyle maven integration

To run scalastyle on your code:

```
$ mvn scalastyle:check
```

or to run testing + scalastyle:

```
$ mvn verify
```

VincenzoFerme commented on benchflow-dsl/pom.xml:
```diff
@@ -175,6 +175,30 @@
+                <sourceDirectory>${basedir}/src/main/scala</sourceDirectory>
+                <testSourceDirectory>${basedir}/src/test/scala</testSourceDirectory>
+                <configLocation>${basedir}/../scalastyle-config.xml</configLocation>
+                <outputFile>${project.basedir}/scalastyle-output.xml</outputFile>
```
@igormoreno you might need to add this file in `.gitignore`

jeppe-style commented:
@VincenzoFerme @igormoreno in reference to https://github.com/benchflow/benchflow/pull/67#pullrequestreview-28423957
- the issue (https://github.com/benchflow/benchflow/issues/58) describing what we should change to replace `var` with `val` (example is in the code)
- the issue (https://github.com/benchflow/benchflow/issues/106) describing what tests are missing

jeppe-style commented:
I worked on the code review. Most of the things are regarding indentation or things we had already talked about changing, or code in the package `simone`. Most of the severe issues are incorporated in a new commit (below).

- [YamlErrorHandler - Call-by-name parameter should be in last position](https://www.codacy.com/app/BenchFlow/benchflow/file/5831089621/issues/source?fileBranchId=4326789#l@result.result.startLine). It says to change the order of the parameters. It would mean changes in ~50 places. Ok to ignore?
- [Time - Covariant equals without overriding equals(java.lang.Object).](https://www.codacy.com/app/BenchFlow/benchflow/file/5831089654/issues/source?bid=4283320&fileBranchId=4326789#l18). This seems to be not correct since the parameter is `scala.Any`

jeppe-style committed:
DSL: Fixes issues from Codacy. Removes unused classes.

jeppe-style committed:
Merge branch 'devel' of https://github.com/benchflow/benchflow into feature-dsl-v1-initial

* 'devel' of https://github.com/benchflow/benchflow:
  Minor changes to adapt to style guidelines
  Remove unjustified vars
  Updates details about where we collect which notes
  Adds tests folder with DSL examples.
  Update README.md
  Update README.md
  Update README.md

jeppe-style committed:
Updates load bechflow-test according to DSL. Updates DSL tests to use the file in the new location.

jeppe-style commented:
We are merging! 
@VincenzoFerme please carefully check that all the remaining work is mentioned in issues, thanks
